src/bigistore.c

```diff
@@ -1099,6 +1099,58 @@ Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
     PG_RETURN_POINTER(aout);
 }
 
+static void bigistore_clamp_pass(BigIStore * is, int32 end_key, int delta_dir)
+{
+    BigIStorePair * pairs;
+    int64 result = 0;
+    int   index  = 0, count = 0;
+
+    pairs = FIRST_PAIR(is, BigIStorePair);
+    index = delta_dir > 0 ? 0 : is->len - 1;
+    while ( ((delta_dir > 0) && (index < is->len && pairs[index].key <= end_key)) ||
+            ((delta_dir < 0) && (index >= 0      && pairs[index].key >= end_key)) )
+    {
+        is->buflen -= bigis_pair_buf_len(pairs + index);
+        result = DirectFunctionCall2(int8pl, result, pairs[index].val);
+        index += delta_dir, count++;
+    }
+
+    if (count) {
+        /* back to the last element that is to be clamped */
+        index -= delta_dir, count--;
+
+        /* put the sum result in its place */
+        pairs[index].key = end_key;
+        pairs[index].val = result;
+
+        /* truncate the rest */
+        is->len -= count;
+        is->buflen += bigis_pair_buf_len(pairs + index);
+        if (delta_dir > 0)
+            memmove(pairs, pairs + index, is->len * sizeof(BigIStorePair));
+    }
+
+    SET_VARSIZE(is, ISHDRSZ + (is->len * sizeof(BigIStorePair)));
+}
+
+PG_FUNCTION_INFO_V1(bigistore_clamp_below);
+Datum bigistore_clamp_below(PG_FUNCTION_ARGS)
+{
+    BigIStore * is      = PG_GETARG_BIGIS_COPY(0);
+    int32       end_key = PG_GETARG_INT32(1);
+    bigistore_clamp_pass(is, end_key, 1);
+    PG_RETURN_POINTER(is);
+}
+
+PG_FUNCTION_INFO_V1(bigistore_clamp_above);
+Datum bigistore_clamp_above(PG_FUNCTION_ARGS)
+{
+    BigIStore * is      = PG_GETARG_BIGIS_COPY(0);
```

Comment:
This implementation optimizes for the clamp case. Maybe it's worth checking the performance benefit / penalty 
of not getting a copy and only creating a new istore in the case there is actually something to clamp

Reply:
Very good point.